pinot-query-planner/src/main/java/org/apache/pinot/query/planner/nodes/RexExpression.java

```diff
@@ -0,0 +1,157 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.planner.nodes;
+
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.util.NlsString;
+import org.apache.pinot.query.planner.nodes.serde.ProtoProperties;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+
+/**
+ * {@code RexExpression} is the serializable format of the {@link RexNode}.
+ */
+public abstract class RexExpression {
+  @ProtoProperties
+  protected SqlKind _sqlKind;
+  @ProtoProperties
+  protected RelDataType _dataType;
+
+  public SqlKind getKind() {
+    return _sqlKind;
+  }
+
+  public RelDataType getDataType() {
+    return _dataType;
+  }
+
+  public static RexExpression toRexExpression(RexNode rexNode) {
+    if (rexNode instanceof RexInputRef) {
+      return new RexExpression.InputRef(((RexInputRef) rexNode).getIndex());
+    } else if (rexNode instanceof RexLiteral) {
+      RexLiteral rexLiteral = ((RexLiteral) rexNode);
+      return new RexExpression.Literal(rexLiteral.getType(), rexLiteral.getTypeName(), rexLiteral.getValue());
+    } else if (rexNode instanceof RexCall) {
+      RexCall rexCall = (RexCall) rexNode;
+      List<RexExpression> operands = rexCall.getOperands().stream().map(RexExpression::toRexExpression)
+          .collect(Collectors.toList());
+      return new RexExpression.FunctionCall(rexCall.getKind(), rexCall.getType(), rexCall.getOperator().getName(),
+          operands);
+    } else {
+      throw new IllegalArgumentException("Unsupported RexNode type with SqlKind: " + rexNode.getKind());
+    }
+  }
+
+  private static Comparable convertLiteral(Comparable value, SqlTypeName sqlTypeName, RelDataType dataType) {
+    switch (sqlTypeName) {
+      case BOOLEAN:
+        return (boolean) value;
+      case DECIMAL:
```

Comment:
The docs of `RexLiteral` don't seem to suggest that underlying SqlTypeName for numeric will be wrapped inside DECIMAL

https://calcite.apache.org/javadocAggregate/org/apache/calcite/rex/RexLiteral.html

Reply:
its actually the value that matters. we need to convert the list of Literal value wrap to pinot specific types. 
there's addition things that needs to be addressed, for example a `IN ('a', 'b')` now is encoded as a Sarg('a', 'b') where each string is actually an NlsString. so there are more things to deal with. 

will follow up with a separate PR on the literal conversion